### PR TITLE
Add new `argument` vim text object

### PR DESCRIPTION
**Pull request by vultix**
This PR adds a new `argument` vim text object, inspired by [targets.vim](https://github.com/wellle/targets.vim).

As it's the first vim text object to use the syntax tree, it needed to operate on the `Buffer` level, not the `MultiBuffer` level, then map the buffer coordinates to `DisplayPoint` as necessary.

This required two main changes:
1. `innermost_enclosing_bracket_ranges` and `enclosing_bracket_ranges` were moved into `Buffer`.  The `MultiBuffer` implementations were updated to map to/from these.
2. `MultiBuffer::excerpt_containing` was made public, returning a new `MultiBufferExcerpt` type that contains a reference to the excerpt and methods for mapping to/from `Buffer` and `MultiBuffer` offsets and ranges.

Release Notes:
- Added new `argument` vim text object, inspired by [targets.vim](https://github.com/wellle/targets.vim).

### CODE REVIEW

**Comment by vultix** [comment]
@cla-bot check

**Comment by ConradIrwin** [comment]
@vultix thanks for taking a pass at this! If you'd like to work with me to get this over the line feel free to book time: https://calendly.com/conradirwin/pairing.

Otherwise, I'll leave comments and you can address in your own time.

**Comment by ConradIrwin** [comment]
* I'm not sure about a better way to filter out strings, @maxbrunsfeld may know if we can use the tree-sitter stuff (it also seems like some languages have their comment characters defined as brackets).
* For `,` as a separator, this seems like a reasonable place to start; we may also want `;` too for things like C-style for loops, but I'm happy either way.
* I think it's right to use the syntax-tree and not characters to do this (though it can lead to some disappointment if the language isn't supported; when the language is supported it works much better).
* To get the right buffer, you can use text_anchor_for_position on the multi buffer. I'm less sure about the number of syntax layers. Looking at [syntax_ancestor_for_node](https://github.com/zed-industries/zed/blob/1459e7d00adf84551f731e01e35f35a620fdcb06/crates/language/src/buffer.rs#L2599) (option-up) it iterates through them all; that said, I can't imagine you'd likely need to handle more than one for just a function argument.

File: crates/vim/src/object.rs [hunk 1]
```diff
@@ -308,6 +315,157 @@ fn around_next_word(
...
+                    }
+                }
+                break;
+            } else if next.end_byte() > end {
+                end = next.end_byte();
+            }
+        }
+
+        Some(start..end)
+    }
```
Comment (vultix) [inline crates/vim/src/object.rs:457]: I'm still questioning whether using syntax is the best option for this text object. It has quite a few drawbacks compared to a more simple approach that simply matches on brackets and commas:
- `enclosing_bracket_ranges` also returns enclosing open & close strings, which have to be excluded, meaning I manually filter to brackets matching `'(' | '[' | '{' | '<' | '|'` anyways
- The implementation that walks the syntax tree is very complex, around 100 lines of code, to handle various edge cases
- It's not guaranteed that an argument is a single Node in the syntax tree, such as with Rust's macros.  This means that I'm still manually seeking to the nearest comma or bracket character
- I'm struggling to think of a single use case where using the syntax tree gives a better result than using the more naive search

**Comment by vultix** [comment]
@ConradIrwin I believe this is ready for review now.  It ended up being a larger change than I was hoping, but I'm very happy with the simplifications `MultiBufferExcerpt` brings, both to existing and new code.

I'd be happy to hop on a paired programming session if you have questions/comments. Thanks!

File: crates/vim/src/object.rs (hunk 1 above)
Comment (vultix) [reply crates/vim/src/object.rs:457]: The best of both worlds may be attempting the syntax-based selection, then falling back to a text-based one

File: crates/language/src/buffer.rs [hunk 2]
```diff
@@ -2886,6 +2886,52 @@ impl BufferSnapshot {
...
+                if len > existing_len {
+                    continue;
+                }
+            }
+
+            result = Some((open, close));
+        }
+
+        result
+    }
```
Comment (vultix) [inline crates/language/src/buffer.rs:2933]: These two functions were copied pretty much verbatim from the `MultiBuffer` implementations.  The `MultiBuffer` implementations now rely on these

**Pull request merged**